xdp-dump/xdpdump.c

```diff
@@ -385,6 +392,85 @@ static bool capture_on_legacy_interface(struct dumpopt *cfg)
 	return rc;
 }
 
+/*****************************************************************************
+ * guess_full_main_function()
+ *****************************************************************************/
+static char *guess_full_main_function(struct bpf_prog_info *info,
+				      char *function_override)
+{
+	int         nr_types;
+	int         matches = 0;
+	char       *func_name_ptr = NULL;
+	struct btf *btf = NULL;
+
+	/*
+	 * If the function name is longer than 15 characters the bpf_prog_info
+	 * structure will truncate it to 15 characters. Here we try to see if
+	 * we need to check BTF information to figure out the full function
+	 * name.
```

Comment:
Since you already implemented the looping through names, why not just do it unconditionally? That would effectively give us 'prefix matching' so the user only needs to specify the shortest unambiguous function name prefix - kinda like getopt does with long opts...

Reply:
Decided not to do this, as this is the ambiguity we try to workaround introduced in the kernel. In this case, I rather have them explicitly specify the full name!

Reply:
But since you just print out the list if there's any ambiguity that would still happen whether the ambiguity is caused by the kernel length limit or because someone just entered too short a name? And in both cases we'd ask for a longer name? So what's the harm in saving a bit of typing if there is no ambiguity? Or am I missing something else here?

Reply:
Your suggestion would work, but it might hurt lazy people (like me) in the long run. And anyway if I need to type 16+ chars, I would probably cut/paste the function name :)

I'm quite sure (I will at least) script around xdpdump to get traces in a specific use case, now let's assume it all worked because I used a short name that was unique enough, i.e. xdp_test_prog_with_a_long_name_ with the following in the file:

```
# bpftool btf dump id 1072 | grep xdp_test_prog_w
[7] FUNC 'xdp_test_prog_with_a_long_name' type_id=5
[9] FUNC 'xdp_test_prog_with_a_long_name_too' type_id=8
```

Now a new function xdp_test_prog_with_a_long_name_for_testing, now this will result in my scripts to fail. Not having the auto-expansion will protect people like me from this. In addition, if I sent people output of the xdpdump command line, they understand what I was (they need to) capturing independent of the version of the XDP code running.

In addition in the future if we support multiprogram's it can become even more ambiguous as they could supply a command line like: `-p a,b,er,c`. And I have no idea what programs they are tracing if they supply the command line output to support (guess if they have pcapng we could now from the trace).

Reply:
Heh, your argument basically boils down to "let's not add this convenience feature, because users are are just going to do stupid things with it", doesn't it? ;)

I mean, if you're worried that your scripts are going to fail because you didn't provide the full name, just...make sure you provide the full name?

Not going to insist if you're really against this, though, just thought it was an obvious convenience feature that you had basically already implemented :)

Reply:
> Heh, your argument basically boils down to "let's not add this convenience feature, because users are are just going to do stupid things with it", doesn't it? ;)

Yes, this is correct, and the reason why I'm against it, in this case, is that this tool is mainly used to troubleshoot problems, i.e. used by customer support asking operators who might not have the technical knowledge to gather information on their behave. Now let's assume tech support tested something in their lab, and want "xdpdump -i eth -p xdp_test_prog_with_a_long_name_" (yes the argument can be you should not do this, but my experience is that this will not happen in real life as humans are lazy by nature). Now the operator executes this in his production environment and gets the result this is not possible and needs to pick one more specific function... Here you lose another day (assuming troubleshooting happens in off-hours).

Let's keep the full name for now, and see if we ever get feedback from the field... We can always add it later (removing might be harder ;)

Reply:
Heh, fair enough. BTW, how about printing out the actual function being traced at the beginning of the execution, to aid with something like this? Similar to how tcpdump starts out printing this: `listening on eth1, link-type EN10MB (Ethernet), capture size 262144 bytes`

Reply:
Great idea, done in separate commit :)